src/lib/ytdl/YtdlParser.cxx

```diff
@@ -0,0 +1,325 @@
+				context->builder->AddItem(TAG_ARTIST, str.c_str());
+			}
+			break;
+		case YTDL_YAJL_STATE_CREATOR:
+			context->builder->RemoveType(TAG_ARTIST);
+			context->builder->AddItem(TAG_ARTIST, str.c_str());
+			break;
+		case YTDL_YAJL_STATE_DESCRIPTION: context->builder->AddItem(TAG_COMMENT, str.c_str()); break;
+		case YTDL_YAJL_STATE_PLAYLIST_TITLE: context->builder->AddItem(TAG_ALBUM, str.c_str()); break;
+		case YTDL_YAJL_STATE_WEBPAGE_URL: context->webpage_url = str; break;
+		case YTDL_YAJL_STATE_EXTRACTOR: context->extractor = str; break;
+		case YTDL_YAJL_STATE_TYPE: context->type = str; break;
+		case YTDL_YAJL_STATE_URL: context->url = str; break;
+		case YTDL_YAJL_STATE_HEADERS:
+			context->headers[context->header_index].second = str;
+			return 1;
+		case YTDL_YAJL_STATE_ENTRIES:
+			if (context->entry_ptr) {
+				ytdl_yajl_string(context->entry_ptr, value, len);
+				return 1;
+			} else {
+				return 0;
+			}
+		default: break;
+	}
+	context->state = YTDL_YAJL_STATE_NONE;
+
+	return 1;
+}
+
+int YtdlParseContext::ytdl_yajl_integer(void *ctx, long long int value) {
+	auto context = (YtdlParseContext*)ctx;
+	switch (context->state) {
+		case YTDL_YAJL_STATE_DURATION: context->builder->SetDuration(SignedSongTime::FromS((int)value)); break;
+		case YTDL_YAJL_STATE_PLAYLIST_INDEX: context->playlist_index = (int)value; break;
+		case YTDL_YAJL_STATE_ENTRIES:
+			if (context->entry_ptr) {
+				ytdl_yajl_integer(context->entry_ptr, value);
+				return 1;
+			} else {
+				return 0;
+			}
+		default: break;
+	}
+	context->state = YTDL_YAJL_STATE_NONE;
+
+	return 1;
+}
+
+int YtdlParseContext::ytdl_yajl_double(void *ctx, double value) {
+	auto context = (YtdlParseContext*)ctx;
+	switch (context->state) {
+		case YTDL_YAJL_STATE_DURATION: context->builder->SetDuration(SignedSongTime::FromMS((int)(value * 1000))); break;
+		case YTDL_YAJL_STATE_ENTRIES:
+			if (context->entry_ptr) {
+				ytdl_yajl_double(context->entry_ptr, value);
+				return 1;
+			} else {
+				return 0;
+			}
+		default: return ytdl_yajl_integer(ctx, (long long int)value);
+	}
+	context->state = YTDL_YAJL_STATE_NONE;
+
+	return 1;
+}
+
+const void* YtdlParseContext::YajlCallbacks() {
+	static constexpr yajl_callbacks ytdl_yajl_callbacks = {
+		nullptr,
+		nullptr,
+		YtdlParseContext::ytdl_yajl_integer,
+		YtdlParseContext::ytdl_yajl_double,
+		nullptr,
+		YtdlParseContext::ytdl_yajl_string,
+		YtdlParseContext::ytdl_yajl_start_map,
+		YtdlParseContext::ytdl_yajl_map_key,
+		YtdlParseContext::ytdl_yajl_end_map,
+		YtdlParseContext::ytdl_yajl_start_array,
+		YtdlParseContext::ytdl_yajl_end_array,
+	};
+
+	return &ytdl_yajl_callbacks;
+}
+
+bool YtdlParseJson(YtdlParseContext *context, const char *url, YTDL_PLAYLIST_MODE playlist_mode) {
+	yajl_handle handle = nullptr;
+
+	int pid = 0;
+	int pipefd[2] = {-1, -1};
+	bool result = false;
+	int ret;
+	yajl_status status;
+	uint8_t buffer[0x80];
+
+	if (pipe(pipefd) < 0) {
+		goto fail;
+	}
+
+	pid = fork();
```

Comment:
Hm, what is a portable alternative here, is pread available? Or an explicit platform ifdef and use windows APIs?

Reply:
You mean `popen()`? (`pread()` is a different beast)
Yes, that is portable, and in use by the known-bad `pipe` output plugin, but is unsafe, because it doesn't restore the signal block mask and the signal handlers. (Your code doesn't do that either.)
Spawning processes is something I like to avoid completely, because it is extremely difficult to get right.
For your plugin, I think it would be acceptable to disable it (for now) on operating systems which don't support `fork()` (i.e. Windows). But in any case, your code must be proper - i.e. get the thing right, be exception-safe, not break the build and not introduce resource leaks (e.g. fd leaks).

Reply:
Er sorry yeah - I had a popen implementation there initially but swapped it out for manual fork/exec because it bothered me. What you've said makes sense to me.